internal/app/fullscreen/fullscreen.go

```diff
@@ -0,0 +1,204 @@
+	input.Placeholder = "Query"
+	input.Focus()
+
+	model := Model{
+		qm: manager.New(
+			ctx,
+			x.Engine,
+			manager.WithSingleOpts(x.SingleOpts),
+			manager.WithRangeOpts(x.RangeOpts),
+			manager.WithWrite(x.Write)),
+
+		log:  x.Log,
+		mode: Input,
+
+		border: Border{
+			results: lip.NewStyle().
+				Border(lip.RoundedBorder()).
+				Padding(0, 1),
+
+			input: lip.NewStyle().
+				Border(lip.RoundedBorder()).
+				Padding(0, 1),
+		},
+		results: results.New(x.Format),
+		input:   input,
+	}
+
+	_, err := tea.NewProgram(
+		model,
+		tea.WithContext(ctx),
+		tea.WithOutput(x.Out),
+		tea.WithAltScreen(),
+	).Run()
+	return err
+}
+
+type Model struct {
+	qm     manager.QueryManager
+	log    zerolog.Logger
+	mode   Mode
+	latest time.Time
+
+	border  Border
+	results results.Model
+	input   textinput.Model
+}
+
+type Mode int
+
+const (
+	Input Mode = iota
+	Scroll
+)
+
+type Border struct {
+	results lip.Style
+	input   lip.Style
+}
+
+func (x Model) Init() tea.Cmd {
+	return textinput.Blink
+}
+
+func (x Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	if _, ok := msg.(cursor.BlinkMsg); !ok {
+		x.log.Log().Msgf("msg: %T %v", msg, msg)
+	}
+
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		switch msg.Type {
+		case tea.KeyCtrlC:
+			return x, tea.Quit
+
+		case tea.KeyEnter:
+			return x, x.qm.Query(x.input.Value())
+
+		case tea.KeyRunes:
+			if msg.String() == "i" && x.mode == Scroll {
+				x.mode = Input
+				x.input.Focus()
+				return x, textinput.Blink
+			}
+
+		case tea.KeyEscape:
+			if x.mode == Input {
+				x.mode = Scroll
+				x.input.Blur()
+				return x, nil
+			}
+		}
+
+	case manager.AsyncQueryMsg:
+		if x.latest.After(msg.StartedAt) {
+			return x, nil
+		}
+		if x.latest.Before(msg.StartedAt) {
+			x.results.Reset()
+			x.latest = msg.StartedAt
+		}
+
+		buf, done := msg.Buffer.Get()
+		x.results.PushMany(buf)
+		if !done {
+			return x, tea.Tick(50*time.Millisecond, func(_ time.Time) tea.Msg {
+				return msg
+			})
+		}
+		return x, nil
+
+	case error, string, keyval.KeyValue:
+		x.results.Reset()
+		x.results.Push(msg)
+		return x, nil
+
+	case tea.WindowSizeMsg:
+		return x.updateSize(msg), nil
+	}
+
+	return x.updateChildren(msg)
+}
+
+func (x Model) updateSize(msg tea.WindowSizeMsg) Model {
+	const inputLine = 1
+	const cursorChar = 1
+	inputHeight := x.border.input.GetVerticalFrameSize() + inputLine
+
+	x.border.results.Height(msg.Height - x.border.results.GetVerticalFrameSize() - inputHeight)
+	x.border.results.Width(msg.Width - x.border.results.GetHorizontalFrameSize())
+
+	// TODO: I don't know why this +2 is needed.
+	x.results.Height(x.border.results.GetHeight() - x.border.results.GetVerticalFrameSize() + 2)
+
+	// TODO: I think -2 is due to a bug with how the textinput bubble renders padding.
+	x.input.Width = msg.Width - x.border.input.GetHorizontalFrameSize() - len(x.input.Prompt) - cursorChar - 2
+	x.border.input.Width(msg.Width - x.border.input.GetHorizontalFrameSize())
+
+	return x
+}
```

Comment:
We should figure out why the `+2` and `-2` are required here.

Reply:
I simplified the logic around the height of the results area. As for the text input, I decided not to resolve this. I'm going to be rewriting this input from scratch anyhow, so spending more time debugging the usage of this stock input is not worth my time.